etor/src/main/java/gov/hhs/cdc/trustedintermediary/external/database/PostgresDao.java

```diff
@@ -29,62 +29,63 @@ public static PostgresDao getInstance() {
     }
 
     @Override
-    public void upsertMetadata(
-            String receivedSubmissionId,
-            String sentSubmissionId,
-            String sender,
-            String receiver,
-            String hash,
-            Instant timeReceived,
-            Instant timeDelivered,
-            PartnerMetadataStatus deliveryStatus,
-            String failureReason,
-            PartnerMetadataMessageType messageType)
+    public void upsertData(String tableName, List<DbColumn> values, String conflictColumnName)
             throws SQLException {
-
-        try (Connection conn = connectionPool.getConnection();
-                PreparedStatement statement =
-                        conn.prepareStatement(
-                                """
-                                INSERT INTO metadata VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
-                                ON CONFLICT (received_message_id) DO UPDATE SET receiver = EXCLUDED.receiver, sent_message_id = EXCLUDED.sent_message_id, time_delivered = EXCLUDED.time_delivered, delivery_status = EXCLUDED.delivery_status, failure_reason = EXCLUDED.failure_reason
-                                """)) {
-
-            statement.setString(1, receivedSubmissionId);
-            statement.setString(2, sentSubmissionId);
-            statement.setString(3, sender);
-            statement.setString(4, receiver);
-            statement.setString(5, hash);
-
-            Timestamp timestampReceived = null;
-            if (timeReceived != null) {
-                timestampReceived = Timestamp.from(timeReceived);
-            }
-            statement.setTimestamp(6, timestampReceived);
-
-            Timestamp timestampDelivered = null;
-
-            if (timeDelivered != null) {
-                timestampDelivered = Timestamp.from(timeDelivered);
-            }
-            statement.setTimestamp(7, timestampDelivered);
-
-            String deliveryStatusString = null;
-            if (deliveryStatus != null) {
-                deliveryStatusString = deliveryStatus.toString();
+        StringBuilder sqlStatementBuilder =
+                new StringBuilder("INSERT INTO ").append(tableName).append(" (");
+
+        values.forEach(dbColumn -> sqlStatementBuilder.append(dbColumn.name()).append(", "));
+        sqlStatementBuilder.delete(
```

Comment:
This is minor, but since this way of removing the last `,` is used in multiple places, we could have a private method to do this

Reply:
My only concern with this approach is that it's harder to read what the query is doing because it's more fragmented

Reply:
I like this idea of having the private method.  I think I'll also add a code comment that gives an example SQL statement that helps us understand what is being constructed.